File: lua/tuning.lua
local Notation = require("notation")
local log = require("log")
local scales = require("default.scales")
local tuning_presets = require("default.tuning_presets")

local tuning = {}

tuning.snap_labels = { "Diatonic", "Chromatic", "Fine" }

local function unit(index)
	local v = tuning.new_interval()
	v[index] = 1
	return v
end

local function pad(p)
	local new = {}
	for i = 1, tuning.rank do
		new[i] = p[i] or 0
	end
	return new
end

-- Load a tuning from definition table
function tuning.load(def)
	local ok, system = pcall(tessera.tuning.new, def, def.notation, scales.candidates)
	if not ok then
		log.error(system)
		return false
	end

	def.notation = system:choice()
	tuning.system = system

	-- number of coordinates in a note
	tuning.rank = system:len()

	-- size in semitones of each coordinate
	tuning.generator_pitches = system:generator_pitches()

	tuning.notation = Notation.new(system:style())

	-- interval definitions
	tuning.octave = pad({ 1 })
	tuning.tone = pad({ -1, 2 }) -- whole tone
	tuning.semitone = pad({ 3, -5 }) -- diatonic semitone
	tuning.chroma = pad({ -4, 7 }) -- apotome, chromatic semitone
	if tuning.get_relative_pitch(tuning.chroma) < 0 then
		tuning.chroma = tuning.mul(tuning.chroma, -1)
	end

	local accidentals = system:style().accidentals

	-- Small steps for fine editing.
	-- If there are accidentals, they take precedence.
	-- Otherwise in an equal temperament, one step.
	tuning.comma = nil
	tuning.comma_alt = nil
	tuning.comma_alt2 = nil
	tuning.chroma_alt = nil
	if tuning.rank >= 3 then
		tuning.comma = unit(3)
		if tuning.rank >= 4 then
			tuning.comma_alt = unit(4)
		end
		if tuning.rank >= 5 then
			tuning.comma_alt2 = unit(5)
		end

		if #accidentals >= 1 and accidentals[1].ratio == "81/80" then
			-- 25/24 just chromatic semitone (6/5 - 5/4)
			tuning.chroma_alt = pad({ -4, 7, -2 })
		end
		if #accidentals >= 1 and accidentals[1].ratio == "64/63" then
			-- 28/27 septimal minor second (9/8 - 7/6)
			tuning.chroma_alt = pad({ 3, -5, -1 })
		end
	elseif system:step() then
		tuning.comma = pad(system:step())
	else
		-- pythagorean comma C - Dbb
		tuning.comma = pad({ 7, -12 })
		if tuning.get_relative_pitch(tuning.comma) < 0 then
			tuning.comma = tuning.mul(tuning.comma, -1)
		end
	end

	tuning.diatonic = system:scale(1)
	tuning.chromatic = system:scale(2)
	tuning.fine = system:scale(3)

	tuning.tables = { tuning.diatonic, tuning.chromatic, tuning.fine }

	-- Projections from notes to scale indices. Scales without one are looked up by pitch.
	tuning.maps = {}
	for i, t in ipairs(tuning.tables) do
		tuning.maps[t] = system:scale_map(i)
		if not tuning.maps[t] then
			log.warn(tuning.snap_labels[i] .. " scale has no projection, using nearest pitch.")
		end
	end
	tuning.center = pad(tuning.center or {})

	return true
end

-- Set the tuning for the current project.
function tuning.set(def)
	if not tuning.load(def) then
		return false
	end

	local name = def.name or def.subgroup
	log.info("Loading tuning: " .. name .. " (" .. def.notation.accidentals .. " accidentals)")
	project.settings.tuning = util.clone(def)

	-- TODO: do a proper conversion here

	-- fix up the number of coordinates
	for _, ch in ipairs(project.channels) do
		if ch.notes then
			for _, note in ipairs(ch.notes) do
				note.interval = pad(note.interval)
			end
		end
	end
	return true
end

-- Load the tuning from a project that was just loaded or created.
function tuning.load_project()
	if project.settings.tuning then
		if tuning.load(project.settings.tuning) then
			return
		end
		log.error("Failed to load tuning from project, using default.")
	end

	-- fallback to default if nothing works
	project.settings = tuning.load_default()
end

function tuning.load_default()
	local def = tuning_presets.default()
	assert(tuning.load(def), "Failed to load default tuning")
	return def
end

function tuning.new_interval()
	local new = {}
	for i = 1, tuning.rank do
		new[i] = 0
	end
	return new
end

function tuning.get_center(p)
	return tuning.center
end
function tuning.set_center(p)
	tuning.center = util.clone(p)
end

-- Size in semitones of every note in a scale, above the unison. Cached per scale.
local scale_pitches = setmetatable({}, { __mode = "k" })
local function get_scale_pitches(t)
	local sp = scale_pitches[t]
	if not sp then
		sp = {}
		for i, v in ipairs(t) do
			sp[i] = tuning.get_relative_pitch(v)
		end
		scale_pitches[t] = sp
	end
	return sp
end

-- Index of the note in scale t that is closest to a pitch r, in semitones above the unison.
-- The scale repeats every octave and index 0 is the unison.
local function nearest_index(t, r)
	local sp = get_scale_pitches(t)
	local o = tuning.generator_pitches[1]
	local n = #t

	local oct = math.floor(r / o)
	r = r - oct * o

	local best = 0
	local d_best = math.huge
	for i = 1, n do
		local d = math.abs(r - sp[i])
		if d < d_best then
			best = i - 1
			d_best = d
		elseif d > d_best then
			-- sorted, so it only gets worse
			break
		end
	end
	-- the unison in the next octave
	if math.abs(o - r) < d_best then
		best = n
	end
	return oct * n + best
end

-- Given some pitch p, find interval in current grid that is closest
function tuning.snap(p)
	local t = tuning.tables[project.settings.snap_pitch]
	assert(t)
	local rel = p - tuning.get_pitch(tuning.center)
	return tuning.from_table(t, nearest_index(t, rel) + tuning.get_index(t, tuning.center))
end

function tuning.snap_interval(f)
	local t = tuning.tables[project.settings.snap_pitch]
	local steps = tuning.get_index(t, f)
	return tuning.from_table(t, steps)
end

-- Look up interval in table, correcting for octave offsets
function tuning.from_table(t, i)
	local start = tuning.get_index(t, tuning.center)

	i = i - start
	local s = #t
	local oct = math.floor(i / s)
	i = i - oct * s
	local p = util.clone(t[i + 1])
	p[1] = p[1] + oct
	p = tuning.add(p, tuning.center)

	return p
end

-- Indexed by midi number, middle C = midi note number 60.
-- Note: we currently assume #chromatic = 12 so this works.
function tuning.from_midi(n)
	return tuning.from_table(tuning.chromatic, n - 60)
end

local BLACK_KEYS = { [1] = true, [3] = true, [6] = true, [8] = true, [10] = true }

-- build midi note table
function tuning.input_map()
	local rows = {}
	for i = 1, #tuning.chromatic + 1 do
		local midi = 59 + i
		local interval = tuning.from_midi(midi)
		local intervals = tuning.system:simple_spellings(interval)
		local spellings = {}
		for _, v in ipairs(intervals) do
			table.insert(spellings, tuning.get_name(v))
		end
		spellings = util.dedup(spellings)

		local ratios = tuning.system:simple_ratios(interval)
		rows[i] = {
			-- name = tuning.get_name(interval),
			name = table.concat(spellings, " = "),
			ratio = table.concat(ratios, " ~ "),
			black = BLACK_KEYS[(midi - 60) % 12] or false,
		}
	end
	return rows
end

-- Index of interval p in scale t.
-- With a projection this is exact, otherwise it is the note closest in pitch.
function tuning.get_index(t, p)
	local map = tuning.maps[t]
	if map then
		local index = 0
		for i, v in ipairs(map) do
			index = index + v * (p[i] or 0)
		end
		return index
	end
	return nearest_index(t, tuning.get_relative_pitch(p))
end

-- Convert interval to pitch.
function tuning.get_pitch(p)
	return 60 + tuning.get_relative_pitch(p)
end

function tuning.get_relative_pitch(p)
	local f = 0
	for i, v in ipairs(p) do
		f = f + v * tuning.generator_pitches[i]
	end
	return f
end

-- Name of a note in the current tuning.
function tuning.get_name(p)
	if project.settings.relative_note_names then
		p = tuning.sub(p, tuning.center)
	elseif tuning.rank > 2 then
		-- only move by extra accidentals
		local new = {}
		for i = 1, tuning.rank do
			if i > 2 then
				new[i] = p[i] - tuning.center[i]
			else
				new[i] = p[i]
			end
		end
		p = new
	end

	return tuning.notation:name(p)
end

-- basic arithmetic functions

function tuning.add(a, b)
	-- add two pitches a and b
	local new = {}
	for i = 1, tuning.rank do
		new[i] = a[i] + b[i]
	end
	return new
end

function tuning.sub(a, b)
	-- subtract b from a
	local new = {}
	for i = 1, tuning.rank do
		new[i] = a[i] - b[i]
	end
	return new
end

function tuning.mul(a, b)
	-- multiply pitch a by scalar b
	local new = {}
	for i = 1, tuning.rank do
		new[i] = a[i] * b
	end
	return new
end

function tuning.eq(a, b)
	-- check equality for a and b
	for i = 1, tuning.rank do
		if a[i] ~= b[i] then
			return false
		end
	end
	return true
end

return tuning

File: src/tuning.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use xen_utils::{Notation, Subgroup, Temperament};

// Number of fifths in the chain scales, and how far down from the root they start.
const DIATONIC: (usize, i64) = (7, 1);
const CHROMATIC: (usize, i64) = (12, 4);
// Range of sizes for the fine chain of fifths. The largest one that has a projection is used.
const FINE_CHAIN: std::ops::RangeInclusive<usize> = 15..=31;
// How many spellings of a scale note to look through for one that reads as the just ratio.
const SPELLING_OPTIONS: usize = 8;

// Which temperament to use.
// Either give `commas` to temper out, or `et` for an equal temperament.
// If neither is present, it is just intonation.
//
// This is part of the tuning definition in the save file, which also has a name and a NotationChoice.
//
// Note: mlua serializes None as a null value that is truthy in Lua,
// so anything that goes to Lua needs `skip_serializing_if` on its options.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct TemperamentDef {
	// Subgroup in shorthand, e.g. "2.3.5.7"
	pub subgroup: String,
	// Commas to temper out as "p/q" strings.
	#[serde(default)]
	pub commas: Vec<String>,
	// Equal division of the octave.
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub et: Option<i64>,
}

// Which of the notation options of a temperament to use.
//
// An option is identified by its number of accidentals (see Notation::with_count),
// and whether the one that is half an apotome is written as a half sharp.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct NotationChoice {
	pub accidentals: usize,
	#[serde(default)]
	pub half_sharp: bool,
}

// How to draw spellings, for notation.lua.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct NotationStyle {
	// One for every coordinate after the octave and the fifth.
	pub accidentals: Vec<AccidentalStyle>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AccidentalStyle {
	// Just ratio of the accidental.
	pub ratio: Ratio,
	// Written as half a sharp, combined with the sharps and flats.
	pub half_sharp: bool,
}

// A notation that can be picked in the settings.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NotationOption {
	pub choice: NotationChoice,
	pub recommended: bool,
	pub style: NotationStyle,
	// How the primes beyond 3 are written.
	pub spellings: Vec<PrimeSpelling>,
}

// Just intonation scales to try for each kind of scale, in order of preference.
// Each scale is a list of ratios above the unison, up to and including the octave.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ScaleCandidates {
	#[serde(default)]
	pub diatonic: Vec<Vec<String>>,
	#[serde(default)]
	pub chromatic: Vec<Vec<String>>,
	#[serde(default)]
	pub fine: Vec<Vec<String>>,
}

// Notes in one octave, sorted by pitch. The first one is always the unison.
#[derive(Debug, Clone, Default)]
pub struct Scale {
	pub notes: Vec<Vec<i64>>,
	// Linear map from notation coordinates to steps of the scale,
	// that takes every note in the scale to its own index.
	// None if there is no such map, then lookups have to go by pitch.
	pub map: Option<Vec<i64>>,
}

#[derive(Debug)]
pub struct TuningSystem {
	notation: Notation,
	choice: NotationChoice,
	style: NotationStyle,
	// size of each notation coordinate in semitones.
	pitches: Vec<f64>,
	// diatonic, chromatic, fine
	scales: [Scale; 3],
}

type Ratio = (u64, u64);

// How a notation writes one prime, reduced to an octave (5/4, 7/4, 11/8, ..).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PrimeSpelling {
	pub ratio: Ratio,
	// On the nominal just intonation gives it, in notation coordinates.
	// None if the notation can't write it there.
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub nominal: Option<Vec<i64>>,
	// The next best way to write it, if that is worth showing next to the nominal.
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub alternative: Option<Vec<i64>>,
}

impl TuningSystem {
	// Without a notation choice, the recommended one is used.
	pub fn new(
		def: &TemperamentDef,
		choice: Option<NotationChoice>,
		candidates: &ScaleCandidates,
	) -> Result<Self, String> {
		let options = notation_options(&temperament(def)?)?;
		let index = if let Some(choice) = choice {
			options
				.iter()
				.position(|(_, c)| *c == choice)
				.ok_or_else(|| format!("Notation {choice:?} is not available"))?
		} else {
			options.iter().position(|(n, _)| n.keeps_nominals()).unwrap_or(0)
		};
		let (notation, choice) = options.into_iter().nth(index).expect("index is in range");

		let pitches = (0..notation.len())
			.map(|i| {
				notation
					.pitch(&basis_vec(i, notation.len()))
					.map(|cents| cents / 100.0)
					.map_err(err)
			})
			.collect::<Result<Vec<_>, _>>()?;

		let style = notation_style(&notation, choice.half_sharp);
		let mut system = Self { notation, choice, style, pitches, scales: Default::default() };
		let diatonic = system
			.first_candidate(&candidates.diatonic)
			.unwrap_or_else(|| system.chain_scale(DIATONIC.0, DIATONIC.1));
		let chromatic = system
			.first_candidate(&candidates.chromatic)
			.unwrap_or_else(|| system.chain_scale(CHROMATIC.0, CHROMATIC.1));
		let fine = system.fine_scale(&candidates.fine)?;
		system.scales = [diatonic, chromatic, fine];

		Ok(system)
	}

	pub fn choice(&self) -> NotationChoice {
		self.choice
	}

	pub fn style(&self) -> &NotationStyle {
		&self.style
	}

	#[allow(clippy::len_without_is_empty)]
	pub fn len(&self) -> usize {
		self.notation.len()
	}

	// size of each notation generator, semitones
	pub fn pitches(&self) -> &[f64] {
		&self.pitches
	}

	// Scale by index (0 = diatonic, 1 = chromatic, 2 = fine).
	pub fn scale(&self, index: usize) -> &Scale {
		&self.scales[index]
	}

	// One step up of an equal temperament, spelled in the notation.
	// None for higher rank temperaments, or if the notation can't write it.
	pub fn step(&self) -> Option<Vec<i64>> {
		if self.notation.temperament().rank() != 1 {
			return None;
		}
		let step = self.notation.spell(&[1]).ok()?;
		if self.pitch(&step) < 0.0 {
			return Some(step.iter().map(|x| -x).collect());
		}
		Some(step)
	}

	fn octave(&self) -> f64 {
		self.pitches[0]
	}

	fn pitch(&self, note: &[i64]) -> f64 {
		note.iter().zip(&self.pitches).map(|(&n, p)| n as f64 * p).sum()
	}

	// Move a note into the octave above the unison.
	fn reduce(&self, note: &mut [i64]) {
		// Small offset so notes that are octaves apart in the temperament reduce identically
		note[0] -= ((self.pitch(note) + 1e-9) / self.octave()).floor() as i64;
	}

	// Well-formed scale from a chain of fifths, reduced to one octave.
	//
	// If the temperament closes the circle of fifths early (like 15et, where B is C),
	// there would be duplicates. In that case the note closest to the root is kept.
	fn chain_scale(&self, n: usize, offset: i64) -> Scale {
		let mut fifths: Vec<i64> = (0..n as i64).map(|i| i - offset).collect();
		fifths.sort_by_key(|&f| (f.abs(), -f));

		let mut seen = HashSet::new();
		let mut notes = Vec::new();
		for f in fifths {
			let mut note = vec![0; self.len()];
			note[1] = f;
			self.reduce(&mut note);
			if seen.insert(self.temper(&note)) {
				notes.push(note);
			}
		}
		self.sort(&mut notes);
		let map = self.projection(&notes);
		Scale { notes, map }
	}

	// The first candidate that works in this tuning.
	fn first_candidate(&self, candidates: &[Vec<String>]) -> Option<Scale> {
		candidates.iter().find_map(|ratios| self.ji_scale(ratios))
	}

	// Scale from a list of just ratios, spelled in the notation.
	//
	// None if some ratio isn't in the subgroup, if the temperament merges two notes,
	// or if there is no projection.
	fn ji_scale(&self, ratios: &[String]) -> Option<Scale> {
		let subgroup = self.notation.subgroup();
		let octave = basis_vec(0, subgroup.dim());

		let unison = vec![0; self.len()];
		let mut seen = HashSet::from([self.temper(&unison)]);
		let mut notes = vec![unison];
		for ratio in ratios {
			let interval = subgroup.parse_ratio(ratio).ok()?;
			if interval == octave {
				continue;
			}
			let mut note = self.spell_literal(&interval)?;
			self.reduce(&mut note);
			if !seen.insert(self.temper(&note)) {
				return None;
			}
			notes.push(note);
		}
		self.sort(&mut notes);
		let map = self.projection(&notes)?;
		Some(Scale { notes, map: Some(map) })
	}

	// Spell a just interval as it is written literally if possible, like ^Db for 16/15
	// in 41et. Otherwise use the best spelling of its tempered image.
	fn spell_literal(&self, interval: &[i64]) -> Option<Vec<i64>> {
		let mut options = self.notation.spellings_interval(interval, SPELLING_OPTIONS).ok()?;
		let literal = options
			.iter()
			.position(|s| self.notation.to_interval(s).is_ok_and(|i| i == interval));
		Some(options.swap_remove(literal.unwrap_or(0)))
	}

	fn fine_scale(&self, candidates: &[Vec<String>]) -> Result<Scale, String> {
		if self.notation.temperament().rank() > 1 {
			if let Some(scale) = self.first_candidate(candidates) {
				return Ok(scale);
			}
			// Largest chain of fifths that has a projection.
			for n in FINE_CHAIN.rev() {
				let scale = self.chain_scale(n, chain_offset(n));
				if scale.map.is_some() {
					return Ok(scale);
				}
			}
			let n = *FINE_CHAIN.end();
			return Ok(self.chain_scale(n, chain_offset(n)));
		}

		// Equal temperament: every step, spelled by the notation.
		// The temperament mapping itself is the projection.
		let mut map: Vec<i64> = (0..self.len())
			.map(|i| self.temper(&basis_vec(i, self.len()))[0])
			.collect();
		let sign = map[0].signum();
		map.iter_mut().for_each(|x| *x *= sign);

		let notes = (0..map[0])
			.map(|k| self.notation.spell(&[k * sign]).map_err(err))
			.collect::<Result<Vec<_>, _>>()?;
		Ok(Scale { notes, map: Some(map) })
	}

	// Linear map that takes every note of the scale to its index, if there is one.
	//
	// This is the patent val for the size of the scale, applied to the just reading of
	// each notation coordinate. It exists when the scale is a constant structure that
	// the val agrees with.
	fn projection(&self, notes: &[Vec<i64>]) -> Option<Vec<i64>> {
		let n = notes.len() as f64;
		let val: Vec<i64> = self
			.notation
			.subgroup()
			.log_primes()
			.iter()
			.map(|l| (n * l).round() as i64)
			.collect();
		let map: Vec<i64> = (0..self.len())
			.map(|i| {
				let interval = self
					.notation
					.to_interval(&basis_vec(i, self.len()))
					.expect("right length");
				dot(&interval, &val)
			})
			.collect();

		let consistent = notes.iter().enumerate().all(|(i, note)| dot(&map, note) == i as i64);
		consistent.then_some(map)
	}

	fn temper(&self, note: &[i64]) -> Vec<i64> {
		self.notation.temper(note).expect("note has the right length")
	}

	fn sort(&self, notes: &mut [Vec<i64>]) {
		notes.sort_by(|a, b| self.pitch(a).total_cmp(&self.pitch(b)));
	}
}

// Every notation that can be used for a temperament.
pub fn notations(def: &TemperamentDef) -> Result<Vec<NotationOption>, String> {
	let options = notation_options(&temperament(def)?)?;
	Ok(options
		.into_iter()
		.map(|(notation, choice)| NotationOption {
			choice,
			recommended: notation.keeps_nominals(),
			style: notation_style(&notation, choice.half_sharp),
			spellings: prime_spellings(&notation),
		})
		.collect())
}

// The notations xen_utils offers, and for those that can use a half sharp, whether they do.
fn notation_options(temperament: &Temperament) -> Result<Vec<(Notation, NotationChoice)>, String> {
	let mut options = Vec::new();
	for notation in Notation::options(temperament).map_err(err)? {
		let accidentals = notation.len() - 2;
		let plain = NotationChoice { accidentals, half_sharp: false };
		match half_sharp_index(&notation) {
			Some(i) => {
				// 33/32 is always written as a half sharp, others get the choice.
				if generator_ratios(&notation)[i] != (33, 32) {
					options.push((notation.clone(), plain));
				}
				options.push((notation, NotationChoice { accidentals, half_sharp: true }));
			},
			None => options.push((notation, plain)),
		}
	}
	Ok(options)
}

fn notation_style(notation: &Notation, half_sharp: bool) -> NotationStyle {
	let half_sharp_index = if half_sharp { half_sharp_index(notation) } else { None };
	let accidentals = generator_ratios(notation)
		.into_iter()
		.enumerate()
		.skip(2)
		.map(|(i, ratio)| AccidentalStyle { ratio, half_sharp: Some(i) == half_sharp_index })
		.collect();
	NotationStyle { accidentals }
}

fn temperament(def: &TemperamentDef) -> Result<Temperament, String> {
	let subgroup: Subgroup = def.subgroup.parse().map_err(err)?;

	match (&def.et, def.commas.is_empty()) {
		(Some(_), false) => Err("Can't give both `et` and `commas`".to_string()),
		(Some(n), true) => Temperament::equal(*n, &subgroup).map_err(err),
		(None, true) => Temperament::from_ji(&subgroup).map_err(err),
		(None, false) => {
			let commas = def
				.commas
				.iter()
				.map(|s| subgroup.parse_ratio(s).map_err(err))
				.collect::<Result<Vec<_>, String>>()?;
			Temperament::from_commas(&commas, &subgroup).map_err(err)
		},
	}
}

fn err(e: xen_utils::Error) -> String {
	e.to_string()
}

// Coordinate of the accidental that can be written as a half sharp, if there is one:
// two of them are the apotome (7 fifths down 4 octaves) in the temperament.
// By construction this can only appear at one accidental.
fn half_sharp_index(notation: &Notation) -> Option<usize> {
	let mut apotome = vec![0; notation.len()];
	(apotome[0], apotome[1]) = (-4, 7);
	let apotome = notation.temper(&apotome).ok()?;

	(2..notation.len()).find(|&i| {
		let image = notation.temper(&basis_vec(i, notation.len())).expect("right length");
		image.iter().zip(&apotome).all(|(a, b)| 2 * a == *b)
	})
}

// How every prime beyond 3 is written, on its nominal and the way `spell` writes it.
fn prime_spellings(notation: &Notation) -> Vec<PrimeSpelling> {
	let basis = notation.subgroup().basis();
	(2..basis.len())
		.zip(notation.nominal_spellings())
		.map(|(i, nominal)| {
			let prime = basis[i];
			let octaves = i64::from(prime.ilog2());
			let mut interval = vec![0; basis.len()];
			(interval[0], interval[i]) = (-octaves, 1);

			// nominal_spellings writes the prime itself, so bring it down the same octaves.
			let nominal = nominal.map(|mut note| {
				note[0] -= octaves;
				note
			});
			let cost = |note: &[i64]| notation.cost(note).expect("always a valid spelling");
			// The cheapest spelling that isn't the one on the nominal.
			// It earns a place next to the nominal by being easier to read: either it is
			// cheaper, or it needs fewer accidentals, like A# next to vBb in meantone.
			let alternative = notation
				.spellings_interval(&interval, 2)
				.expect("always a valid interval")
				.into_iter()
				.find(|spelled| Some(spelled) != nominal.as_ref())
				.filter(|spelled| match nominal.as_deref() {
					Some(nominal) => {
						marks(spelled) < marks(nominal) || cost(spelled) < cost(nominal)
					},
					None => true,
				});

			PrimeSpelling { ratio: (u64::from(prime), 1 << octaves), nominal, alternative }
		})
		.collect()
}

// How many accidentals a spelling writes, not counting sharps and flats.
fn marks(spelling: &[i64]) -> i64 {
	spelling[2..].iter().map(|c| c.abs()).sum()
}

fn generator_ratios(notation: &Notation) -> Vec<(u64, u64)> {
	let mut generators = Vec::new();
	for g in notation.generators() {
		generators.push(
			notation
				.temperament()
				.subgroup()
				.to_ratio(g)
				.expect("always a valid ratio"),
		);
	}
	generators
}

// Fifths down from the root for a chain of n, so that it is centered around D.
fn chain_offset(n: usize) -> i64 {
	n as i64 / 2 - 2
}

fn dot(a: &[i64], b: &[i64]) -> i64 {
	a.iter().zip(b).map(|(x, y)| x * y).sum()
}

fn basis_vec(i: usize, n: usize) -> Vec<i64> {
	let mut unit = vec![0; n];
	unit[i] = 1;
	unit
}
